CedarJava/src/main/java/com/cedarpolicy/value/EntityUID.java

```diff
@@ -17,118 +17,105 @@
 package com.cedarpolicy.value;
 
 import java.util.Optional;
-import java.util.regex.Matcher;
-import java.util.regex.Pattern;
-import java.util.regex.PatternSyntaxException;
+
+import com.cedarpolicy.serializer.JsonEUID;
 
 /**
  * Represents a Cedar Entity UID. An entity UID contains both the entity type and a unique
  * identifier for the entity formatted as <code>TYPE::"ID"</code>.
  */
-public class EntityUID extends Value {
-
-    private static class EUIDValidator {
-        // Any char except {'\','"'} or escaped slash `\\` or escaped quote `\"` or unicode `\u0000`
-        // We need to escape twice, once for Java and once for the Regex
-        private static final String anyCharExceptSlashOrQuote = "[^\\\"\\\\]";
-        private static final String escapedSlash = "\\\\\\\\";
-        private static final String escapedQuote = "\\\\\\\"";
-        private static final String unicodeEscapedPattern = "\\\\u[A-Fa-f0-9]{4,6}";
-        private static final Pattern entityUIDPattern =
-                Pattern.compile(
-                        "^([A-Za-z_]([A-Za-z0-9_])*::)+\"("
-                                + anyCharExceptSlashOrQuote
-                                + "|"
-                                + escapedSlash
-                                + "|"
-                                + escapedQuote
-                                + "|"
-                                + unicodeEscapedPattern
-                                + ")*\"$");
-        private static final long MAXLENGTH = 1024;
-
-        public static boolean validEntityUID(String id) {
-            if (id == null || id.isEmpty()) return false;
-            id = id.trim();
-            if (id.length() > MAXLENGTH) return false;
-            try {
-                Matcher matcher = entityUIDPattern.matcher(id);
-                return matcher.matches();
-            } catch (PatternSyntaxException ex) {
-                return false;
-            }
-        }
-    }
-
-    /** Entity uid. */
-    private final String euid;
+public final class EntityUID extends Value {
+    private final EntityTypeName type;
+    private final EntityIdentifier id;
+    private final String repr;
 
-    /** Entity uid type. */
-    private final Optional<String> type;
-
-    /** Entity uid type. */
-    private final Optional<String> id;
+    static { 
+        System.load(System.getenv("CEDAR_JAVA_FFI_LIB"));
+    }
 
+    /**
+     * Construct an EntityUID from a tyep name and an id
+     * @param type the Entity Type of this EUID
+     * @param id the id portion of the EUID
+     */
+    public EntityUID(EntityTypeName type, EntityIdentifier id) {
+        this.type = type;
+        this.id = id;
+        this.repr = getEUIDRepr(type, id);
```

Comment:
does this mean we do the native call to get the repr every time we make an EUID? is it needed all the time? or should we be lazy about that?

Reply:
I was trying avoid needing to redo the native call for every call to `toString()` without adding state to the object, but I think I've gone down the rate of pre-mature op, and should just do the native call on every `toString()`